File: java/src/SegmentTreeSimple.java
public class SegmentTreeSimple {

	public static int get(int[] t, int i) {
		return t[i + t.length / 2];
	}

	public static void add(int[] t, int i, int value) {
		i += t.length / 2;
		t[i] += value;
		for (; i > 1; i >>= 1)
			t[i >> 1] = Math.max(t[i], t[i ^ 1]);
	}

	// max[a, b]
	public static int max(int[] t, int a, int b) {
		int res = Integer.MIN_VALUE;
		for (a += t.length / 2, b += t.length / 2; a <= b; a = (a + 1) >> 1, b = (b - 1) >> 1) {
			if ((a & 1) != 0)
				res = Math.max(res, t[a]);
			if ((b & 1) == 0)
				res = Math.max(res, t[b]);
		}
		return res;
	}

	// Usage example
	public static void main(String[] args) {
		int n = 11;
		int[] t = new int[n + n];
		add(t, 1, 2);
		add(t, 2, 1);
		add(t, 1, 5);
		System.out.println(7 == max(t, 0, n - 1));
	}
}

File: scala/src/SegmentTreeSimple.scala
object SegmentTreeSimple {

  def get(t: Array[Int], i: Int): Int = t(i + t.length / 2)

  def add(t: Array[Int], i: Int, value: Int) {
    var j = i + t.length / 2
    t(j) += value

    while (j > 1) {
      t(j >> 1) = math.max(t(j), t(j ^ 1))
      j >>= 1
    }
  }

  def max(t: Array[Int], a: Int, b: Int): Int = {
    var res = Int.MinValue
    var i = a + t.length / 2
    var j = b + t.length / 2
    while (i <= j) {
      if ((i & 1) != 0) res = math.max(res, t(i))
      if ((j & 1) == 0) res = math.max(res, t(j))
      i = (i + 1) >> 1
      j = (j - 1) >> 1
    }
    res
  }

  // Usage example
  def main(args: Array[String]) {
    val n = 10
    val t = new Array[Int](n + n)
    add(t, 0, 1)
    add(t, 9, 2)
    println(2 == max(t, 0, 9))
  }
}
